### ISSUE TITLE
Add crate for std/no_std compatibility layer

### ISSUE THREAD

**Issue by ascjones** [paritytech/ink#157]
From https://github.com/paritytech/ink/pull/143#issuecomment-514669372.

> I really like the direction of separating this functionality out into its own sub-crate, however, I think we can do even better than simply moving the bits over. This could potentially act as a std to no_std + alloc compatibility layer instead of just an alloc shim.

> So all std pieces would always be loaded from within this sub-crate.
For the purpose of usability we should also provide a prelude that we can simply import.
Also maybe the name of this compatibility layer should be different from ink_memory.

> What about ink_std? - That's may be a bit confusing due to ink_core existence though ...
Maybe later we can also split apart ink_core into ink_env and ink_storage.
Also we might want to add an ink_traits for bridging between the different sub-crates. At this point I should probably just open a new issue ...

**Comment by Robbepop** [comment]
Before we start work on this let's please use this thread in order to find a useful design first.
I will write down some thoughts about this sub-crate and its interfaces soon.

**Comment by Robbepop** [comment]
After splitting the `ink_memory` crate out of `ink_core` we should add a link from `ink_core` to it so that contract writers do not have to explicitly depend on both crates but can access `ink_memory` from within `ink_core`.

**Comment by Robbepop** [comment]
Since some time now we have the so-called `ink_prelude` that does exactly what this PR mandates. Closed.
